File: Week 5/Python Codes/TwoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        array = nums
        # create a dictionary to store array elements and their indexes       
        for i in range(len(nums)):
            array[i] = (array[i], i)
        
        left, right = 0, len(array)-1

        array.sort() #sorting the nums array

        while(left < right):
            temp_sum = array[left][0] + array[right][0]
            if (temp_sum == target): #if temp_sum equals target, store the indexes in result
                return min((array[left][1], array[right][1]), (array[right][1], array[left][1]))
            elif (temp_sum > target): #decrease the right pointer and point to smaller number
                right -= 1
            else: #increase the left pointer and point to bigger number
                left += 1
        

File: Week 5/C++ Codes/TwoSum.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& A, int target) {
        vector<pair<int, int>> array;

        // create a vector to store array elements and their indexes       
        for(int i=0; i<A.size(); i++)
            array.push_back(make_pair(A[i], i));
        
        int left = 0, right = array.size()-1;
        
        vector<int> result; // created a result vector to store the result indexes
        sort(array.begin(), array.end()); // sorting the array vector

        while(left < right) {
            int temp_sum = array[left].first + array[right].first;
            if (temp_sum == target) { // if temp_sum equals target, store the indexes in result
                result.push_back(array[left].second);
                result.push_back(array[right].second);
                break;
            }
            else if (temp_sum > target) // decrease the right pointer and point to smaller number
                right--;
            else // increase the left pointer and point to bigger number
                left++;
        }
        return result;
    }
};